---
title: "Annotate S288C UTRs from Pelechano 2013 data"
author: "Edward Wallace, Edward.Wallace@ed.ac.uk"
date: "24 March 2020"
output: 
    html_document:
        toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,cache=TRUE)

library(tidyverse)
library(here)
```

# Functions to read, write, manipulate gffs

```{r read_functions}
read_gff <- function(file){
    # tidyverse read gff function from rmonad vignette
    # https://cran.r-project.org/web/packages/rmonad/vignettes/gff-processing.html
    
    readr::read_tsv(
        file,
        col_names = c(
            "Seqid",
            "Source",
            "Type",
            "Start",
            "End",
            "Score",
            "Strand",
            "Phase",
            "Attr"
        ),
        na        = ".",
        comment   = "#",
        col_types = "ccciidcic"
    )
}

write_gff <- function(x,path,...) {
    x %>%
        dplyr::ungroup() %>%
        dplyr::select(Seqid,Source,Type,Start,End,Score,Strand,Phase,Attr) %>%
        tidyr::replace_na(list(Score=".",Strand=".",Phase=".")) %>%
        readr::write_tsv(path=path,...)
}

gene_from_attr <- function(attr_string,
                           before="Name=",gene_str="[A-Z0-9-]+",after="_") {
    # pull a gene name from an attribute string
    match_str = paste0(before,gene_str,after)
    out_strs <- attr_string %>%
        str_extract(match_str) 
    if (nchar(before) > 0 ){
        out_strs <- str_remove(out_strs,before)
    }
    if (nchar(after) > 0 ){
        out_strs <- str_remove(out_strs,after)
    }
    return(out_strs)
}

# test_bit <- gff_CDS$attr[1]
# gene_from_attr(test_bit)

assign_5UTR <- function(gff_join_row,defaultlength=25) {
    Attr.tx <- pull(gff_join_row,Attr.tx)
    if( !is.na(Attr.tx) ) {
        Attr.UTR <- Attr.tx %>%
            str_extract("ID=[\\w-_]+") %>%
            str_remove("ID=") %>% 
            paste0("ID=",.,"_5UTR;Parent=",.) 
    } else {
        Attr.UTR <- pull(gff_join_row,Gene) %>%
            paste0("ID=",.,"_",defaultlength,"nt_5UTR;Parent=",.,"_tx0") 
    }
    Strand <- pull(gff_join_row,Strand)
    if( Strand == "+" ) {
        Start.UTR <- pull(gff_join_row,Start.tx)
        if( is.na(Start.UTR) ) {
            Start.UTR <- pull(gff_join_row,Start.CDS) - defaultlength
        }
        End.UTR <- pull(gff_join_row,Start.CDS) - 1
    } else if( Strand == "-" ) {
        Start.UTR <- pull(gff_join_row,End.CDS) + 1
        End.UTR <- pull(gff_join_row,End.tx)
        if( is.na(End.UTR) ) {
            End.UTR <- pull(gff_join_row,End.CDS) + defaultlength
        }
    }
    tibble( Gene=pull(gff_join_row,Gene),
            Seqid=pull(gff_join_row,Seqid),
            Source=pull(gff_join_row,Source),
            Type="five_prime_UTR",
            Start=Start.UTR,
            End=End.UTR,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=Attr.UTR)
}

assign_3UTR <- function(gff_join_row,defaultlength=125) {
    Attr.tx <- pull(gff_join_row,Attr.tx)
    if( !is.na(Attr.tx) ) {
        Attr.UTR <- Attr.tx %>%
            str_extract("ID=[\\w-_]+") %>%
            str_remove("ID=") %>% 
            paste0("ID=",.,"_3UTR;Parent=",.) 
    } else {
        Attr.UTR <- pull(gff_join_row,Gene) %>%
            paste0("ID=",.,"_",defaultlength,"nt_3UTR;Parent=",.,"_tx0")
    }
    Strand <- pull(gff_join_row,Strand)
    if( Strand == "-" ) {
        Start.UTR <- pull(gff_join_row,Start.tx)
        if( is.na(Start.UTR) ) {
            Start.UTR <- pull(gff_join_row,Start.CDS) - defaultlength
        }
        End.UTR <- pull(gff_join_row,Start.CDS) - 1
    } else if( Strand == "+" ) {
        Start.UTR <- pull(gff_join_row,End.CDS) + 1
        End.UTR <- pull(gff_join_row,End.tx)
        if( is.na(End.UTR) ) {
            End.UTR <- pull(gff_join_row,End.CDS) + defaultlength
        }
    }
    tibble( Gene=pull(gff_join_row,Gene),
            Seqid=pull(gff_join_row,Seqid),
            Source=pull(gff_join_row,Source),
            Type="three_prime_UTR",
            Start=Start.UTR,
            End=End.UTR,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=Attr.UTR)
}

extend_feature <- function(gff_base,IDextl="ID=",IDextr="_ext",
                           ext5prime=25,ext3prime=125,extunstranded=0,
                           type="extension") {
    # extends a stranded feature in a gff_base by defined amounts
    # assumes there is a "Gene" column for ID, not ideal. 
    # This is very slow, maybe worth profiling?
    Strand <- pull(gff_base,Strand)
    if( Strand == "+" ) {
        Start <- pull(gff_base,Start) - ext5prime
        End   <- pull(gff_base,End)   + ext3prime
    } else if( Strand == "-" ) {
        Start <- pull(gff_base,Start) - ext3prime
        End   <- pull(gff_base,End)   + ext5prime
    } else {
        Start <- pull(gff_base,Start) - extunstranded
        End   <- pull(gff_base,End)   + extunstranded
    }
    tibble( Gene=pull(gff_base,Gene),
            Seqid=pull(gff_base,Seqid),
            Source=pull(gff_base,Source),
            Type=type,
            Start=Start,
            End=End,
            Score=NA,
            Strand=Strand,
            Phase=NA,
            Attr=paste0( IDextl, pull(gff_base,Gene), IDextr)
            )
}

union_gff_groups <- function(gff_in,...,type="Union") {
    gff_in %>%
        group_by(...) %>% 
        summarise(
            Seqid=unique(Seqid),
            Source=unique(Source),
            Type=type,
            Start=min(Start),
            End=max(End),
            Score=NA,
            Strand=unique(Strand),
            Phase=NA,
            Attr=unique(Attr)
        )
}

replace_parent <- function(df,niceType="five_prime_UTR") {
    NiceParent <- filter(df, Type==niceType) %>%
        pull(Attr) %>%
        head(n=1) %>%
        str_extract("Parent=[\\w-]+") %>%
        str_remove("Parent=")
    df %>%
        mutate(Attr=
               str_replace(Attr,
                           "Parent=[\\w-]+",
                           paste0("Parent=",NiceParent) )
        ) %>%
        return()
}
```

```{r txgff_functions}
txgff_frompartlength_1 <- function(partlengthdf) {
    L_five_prime_UTR <- partlengthdf$five_prime_UTR
    L_CDS <- partlengthdf$CDS
    L_three_prime_UTR <- partlengthdf$three_prime_UTR
    Parent <- partlengthdf$Parent
    Gene <- partlengthdf$Gene
    txgff1 <- tibble(
        Seqid = rep(Parent,3),
        Type = c("five_prime_UTR",
                 "CDS",
                 "three_prime_UTR"),
        Start = c(1, 
                  L_five_prime_UTR + 1, 
                  L_five_prime_UTR + L_CDS + 1),
        End = c( L_five_prime_UTR, 
                 L_five_prime_UTR + L_CDS,
                 L_five_prime_UTR + L_CDS + L_three_prime_UTR),
        Phase = c( NA, 
                   0,
                   NA),
        Attr = c(paste0("ID=",Parent,"_five_prime_UTR"),
                 paste0("ID=",Gene,"_CDS"),
                 paste0("ID=",Parent,"_three_prime_UTR") )
    )
    return(txgff1)
}
```


# Load data from gffs

These are all from [SGD](https://www.yeastgenome.org/), mostly annotation R64-2-1.

The `longest_full-ORF` and `most_abundant_full-ORF` transcript gffs are based on Pelechano 2013's TIF-seq paper, as updated by SGD and discussed in:
[Transcriptome visualization and data availability at the Saccharomyces Genome Database](https://doi.org/10.1093/nar/gkz892).

```{r load_gffs,dependson="read_functions"}
gff_SGDall <- 
    here::here("data","saccharomyces_cerevisiae_R64-2-1_20150113.gff") %>%
    read_gff()

gff_CDS <- 
    gff_SGDall %>% 
    filter(Type=="CDS") %>%
    mutate(Gene = gene_from_attr(Attr)) 

gff_5pintrons <- 
    gff_SGDall %>% 
    filter(Type=="five_prime_UTR_intron") %>%
    mutate(Gene = gene_from_attr(Attr)) 

gff_txs <- 
    here::here("data","longest_full-ORF_transcripts_ypd.gff3") %>%
    read_gff() %>%
    mutate(Gene = gene_from_attr(Attr,before="ID=")) 

gff_abundtxs <- 
    here::here("data","most_abundant_full-ORF_transcripts_ypd.gff3") %>%
    read_gff() %>%
    mutate(Gene = gene_from_attr(Attr,before="ID=")) 
```

## Which genes have 5'UTR introns? Which have longest annotations?

We need to address 5'UTR introns in particular becase several genes of interest (ribosomal proteins and Ssd1 targets) have 5'UTR introns, and we want to get their sequences right.

```{r check_5pintrons,dependson="load_gffs"}
gff_5pintrons %>%
    dplyr::mutate( longest_tx_annot = ( Gene %in% gff_txs$Gene) ) %>%
    dplyr::select(Gene,longest_tx_annot,Attr) %>%
    print(n=30)
```

```{r write_5pintrons,dependson="load_gffs"}
outfile_5UTR_introns <-
    here::here("out","five_prime_UTR_introns_R64-2-1_20150113.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_5UTR_introns)
write_gff(gff_5pintrons,
          path=outfile_5UTR_introns,
          append=TRUE)
```


Only one gene, YJL130C/URA2, has an annotated 5'UTR intron while not being in the longest_tx list. We can safely get the annotation wrong for this one gene.


# Join CDS annotation

```{r gff_join,dependson="load_gffs"}

# create the primary transcript region 
# only for nuclear verified ORFs
# this should differ only for CDS introns.
gff_verCDSunion <- 
    gff_CDS %>%
    filter( Seqid != "chrmt", str_detect(Attr,"Verified")) %>%
    union_gff_groups(Gene,type="primary_transcript_region")

gff_join_vCDS_txs <- 
    gff_verCDSunion %>%
    left_join(select(gff_txs, Gene, Seqid, Start, End, Strand,Attr),
              by=c("Gene","Seqid","Strand"), suffix=c(".CDS",".tx")) 

gff_join_vCDS_abundtxs <- 
    gff_verCDSunion %>%
    left_join(select(gff_abundtxs, Gene, Seqid, Start, End, Strand,Attr),
              by=c("Gene","Seqid","Strand"), suffix=c(".CDS",".tx")) 
```

# Assign 5' UTRs, although keeping introns

There are 24 annotated 5'UTR introns in R64-2-1 which would ideally be removed from 5'UTRs. First, we keep them.

```{r gff_5UTRs_keepintrons,dependson="gff_join"}
gff_5UTRs_keepintrons <- 
    gff_join_vCDS_txs %>%
    group_by(Gene) %>%
    do(assign_5UTR(.))  %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_5UTRs_keepintrons <- here::here("out","longest_five_prime_UTRs_keepintrons.gff")

write_lines(
    x=c("##gff-version 3", 
        "##five prime UTRs from longest_full-ORF_transcripts_ypd.gff3",
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_5UTRs_keepintrons)

write_gff(x=gff_5UTRs_keepintrons,
          path=outfile_5UTRs_keepintrons,
          append=TRUE)
```

```{r gff_abund5UTRs_keepintrons,dependson="gff_join"}
gff_abund5UTRs_keepintrons <- 
    gff_join_vCDS_abundtxs %>%
    group_by(Gene) %>%
    do(assign_5UTR(.))  %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_abund5UTRs_keepintrons <- here::here("out","abundant_five_prime_UTRs_keepintrons.gff")

write_lines(
    x=c("##gff-version 3", 
        "##five prime UTRs from most_abundant_full-ORF_transcripts_ypd.gff3",
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_abund5UTRs_keepintrons)

write_gff(x=gff_abund5UTRs_keepintrons,
          path=outfile_abund5UTRs_keepintrons,
          append=TRUE)
```

# Assign 3' UTRs

```{r gff_3UTRs,dependson="gff_join"}
gff_3UTRs <- 
    gff_join_vCDS_txs %>%
    group_by(Gene) %>%
    do(assign_3UTR(.)) %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_3UTRs <- here::here("out","longest_three_prime_UTRs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##three prime UTRs from longest_full-ORF_transcripts_ypd.gff3",
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_3UTRs)

write_gff(x=gff_3UTRs,
          path=outfile_3UTRs,
          append=TRUE)
```

```{r gff_abund3UTRs,dependson="gff_join"}
gff_abund3UTRs <- 
    gff_join_vCDS_abundtxs %>%
    group_by(Gene) %>%
    do(assign_3UTR(.)) %>%
    ungroup %>%
    mutate(Source="yeastutrgff")

outfile_abund3UTRs <- here::here("out","abundant_three_prime_UTRs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##three prime UTRs from most_abundant_full-ORF_transcripts_ypd.gff3",
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_abund3UTRs)

write_gff(x=gff_abund3UTRs,
          path=outfile_abund3UTRs,
          append=TRUE)
```

## Remove introns from 5'UTRs using bedtools

```{bash bedtools_subtract_5UTRintrons,eval=FALSE}
bedtools subtract -s \
-a ../out/longest_five_prime_UTRs_keepintrons.gff \
-b ../out/five_prime_UTR_introns_R64-2-1_20150113.gff > \
../out/longest_five_prime_UTRs_nointrons.gff

bedtools subtract -s \
-a ../out/abundant_five_prime_UTRs_keepintrons.gff \
-b ../out/five_prime_UTR_introns_R64-2-1_20150113.gff > \
../out/abundant_five_prime_UTRs_nointrons.gff
```

```{r load_gff_5UTRs_nointrons,dependson=c("gff_5UTRs_keepintrons","bedtools_subtract_5UTRintrons")}
gff_5UTRs_nointrons <- 
    here::here("out","longest_five_prime_UTRs_nointrons.gff") %>%
    read_gff()

gff_abund5UTRs_nointrons <- 
    here::here("out","abundant_five_prime_UTRs_nointrons.gff") %>%
    read_gff()
```

# CDSs with default-length 25nt 5UTRs and 125nt 3'UTRs

```{r gff_CDS_extend,dependson="gff_join"}
gff_verCDSext <- 
    gff_verCDSunion %>%
    group_by(Gene) %>% 
    do(extend_feature(.,ext5prime=25,ext3prime=125,
                      IDextr="_tx0",type="primary_transcript")) %>%
    mutate(Source="yeastutrgff")

outfile_CDSext <- here::here(
    "out",
    "transcripts_fixedlength_25nt5p_125nt3p_UTRs.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_CDSext)

write_gff(x=gff_verCDSext,
          path=outfile_CDSext,
          append=TRUE)
```

## Now make transcripts which are longest if we have data, fixed-length if we don't

```{r gff_longest_or_CDSext,dependson="gff_CDS_extend"}
genes_verified_no_longesttx <- 
    setdiff(gff_verCDSext$Gene,gff_txs$Gene)

gff_longest_or_CDSext <- 
    bind_rows(gff_txs,
              gff_verCDSext %>% 
                  filter(
                     Gene %in% genes_verified_no_longesttx)
              ) %>% 
    arrange(Gene)

outfile_longest_or_CDSext <- here::here(
    "out",
    "longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_longest_or_CDSext)

write_gff(x=gff_longest_or_CDSext,
          path=outfile_longest_or_CDSext,
          append=TRUE)
```

```{r gff_abundant_or_CDSext,dependson="gff_CDS_extend"}
genes_verified_no_abundanttx <- 
    setdiff(gff_verCDSext$Gene,gff_txs$Gene)

gff_abundant_or_CDSext <- 
    bind_rows(gff_abundtxs,
              gff_verCDSext %>% 
                  filter(
                     Gene %in% genes_verified_no_abundanttx)
              ) %>% 
    arrange(Gene)

outfile_abundant_or_CDSext <- here::here(
    "out",
    "abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff")

write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=outfile_abundant_or_CDSext)

write_gff(x=gff_abundant_or_CDSext,
          path=outfile_abundant_or_CDSext,
          append=TRUE)
```

## Remove introns from CDS?

```{r gff_introns,dependson="load_gff"}
gff_allnucmRNAintrons <- 
    gff_SGDall %>% 
    filter(str_detect(Type,"intron"),str_detect(Attr,"Parent=Y")) %>%
    mutate(Gene = gene_from_attr(Attr)) 

outfile_nucmRNA_introns <-
    here::here("out","nuclear_mRNA_introns_R64-2-1_20150113.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##all introns from nuclear chromosome mRNAs, including 5'UTR introns"),
    path=outfile_nucmRNA_introns)
write_gff(gff_allnucmRNAintrons,
          path=outfile_nucmRNA_introns,
          append=TRUE)
```

## Remove introns from longest transcripts using bedtools

```{bash bedtools_subtract_allintrons,eval=FALSE}
bedtools subtract -s \
-a ../out/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff \
-b ../out/nuclear_mRNA_introns_R64-2-1_20150113.gff > \
../out/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.gff

bedtools subtract -s \
-a ../out/abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff \
-b ../out/nuclear_mRNA_introns_R64-2-1_20150113.gff > \
../out/abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts_nointrons.gff
```

# Collected gff with CDS, introns, coding transcripts, UTRs, only

FIXED: Parent of CDS should be the same tx as UTRs.

## Using longest RNA from Pelechano 2013

```{r gff_longestcollected,dependson=c("load_gffs","gff_introns","load_gff_5UTRs_nointrons","gff_3UTRs")}
gff_SGD_chromosomes_ncRNA <-
    gff_SGDall %>%
            filter( Type %in% 
                        c("chromosome",
                          "rRNA_gene","ncRNA_gene","telomerase_RNA_gene",
                          "snoRNA_gene","snRNA_gene","tRNA_gene")
            )

# gff_longestcollected_mRNA %>% filter(Gene=="YFL039C") %>% replace_parent

gff_longestcollected_mRNA <- 
    bind_rows( 
        gff_allnucmRNAintrons  %>%
            mutate(Attr = str_replace(Attr,"Name=","ID=")),
        gff_5UTRs_nointrons %>%
            mutate(Gene=gene_from_attr(Attr,before="ID=")),
        gff_3UTRs %>%
            mutate(Gene=gene_from_attr(Attr,before="ID=")),
        gff_CDS %>% 
            filter(str_detect(Attr,"Verified")) %>%
            mutate(Attr = str_replace(Attr,"Name=","ID="))
    ) %>%
    dplyr::arrange(Seqid,Start,desc(End)) %>%
    group_by(Gene) %>%
    do(replace_parent(.))

gff_longestcollected <- bind_rows(gff_SGD_chromosomes_ncRNA,
                           gff_longestcollected_mRNA) %>%
    dplyr::arrange(Seqid,Start,desc(End))

outfile_gff_longestcollected <-
    here::here("out","gff_Chromosomes_ncRNAs_longestverifiedmRNAparts.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: chromosomes, ncRNA, CDS, 5'UTRs, 3'UTRs, introns"),
    path=outfile_gff_longestcollected)
write_gff(gff_longestcollected,
          path=outfile_gff_longestcollected,
          append=TRUE)


outfile_gff_longestcollected_nochr <-
    here::here("out","gff_ncRNAs_longestverifiedmRNAparts.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: ncRNA, CDS, 5'UTRs, 3'UTRs, introns"),
    path=outfile_gff_longestcollected_nochr)
write_gff(gff_longestcollected %>% filter(Type != "chromosome"),
          path=outfile_gff_longestcollected_nochr,
          append=TRUE)

```

## Using most abundant RNA isoform from Pelechano 2013

```{r gff_abundantcollected,dependson=c("gff_longestcollected")}
gff_abundantcollected_mRNA <- 
    bind_rows( 
        gff_allnucmRNAintrons  %>%
            mutate(Attr = str_replace(Attr,"Name=","ID=")),
        gff_abund5UTRs_nointrons %>%
            mutate(Gene=gene_from_attr(Attr,before="ID=")),
        gff_abund3UTRs %>%
            mutate(Gene=gene_from_attr(Attr,before="ID=")),
        gff_CDS %>% 
            filter(str_detect(Attr,"Verified")) %>%
            mutate(Attr = str_replace(Attr,"Name=","ID="))
    ) %>%
    dplyr::arrange(Seqid,Start,desc(End)) %>%
    group_by(Gene) %>%
    do(replace_parent(.))

gff_abundantcollected <- bind_rows(gff_SGD_chromosomes_ncRNA,
                           gff_abundantcollected_mRNA) %>%
    dplyr::arrange(Seqid,Start,desc(End))

outfile_gff_abundantcollected <-
    here::here("out","gff_Chromosomes_ncRNAs_abundantverifiedmRNAparts.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: chromosomes, ncRNA, CDS, 5'UTRs, 3'UTRs, introns"),
    path=outfile_gff_abundantcollected)
write_gff(gff_abundantcollected,
          path=outfile_gff_abundantcollected,
          append=TRUE)


outfile_gff_abundantcollected_nochr <-
    here::here("out","gff_ncRNAs_abundantverifiedmRNAparts.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: ncRNA, CDS, 5'UTRs, 3'UTRs, introns"),
    path=outfile_gff_abundantcollected_nochr)
write_gff(gff_abundantcollected %>% filter(Type != "chromosome"),
          path=outfile_gff_abundantcollected_nochr,
          append=TRUE)

```

### Awkward rearrangement of Collected gff for input to featureCounts

featureCounts in the subread package will only quantify reads from one feature type at once in a gff. So this workaround (awkward rearrangement) takes the information in the "type" column and distributes int to "attr".

UPDATE: didn't work, featureCounts crashed for an unknown reason.

```{r gff_longestcollected_featureCounts_workaround,dependson="gff_longestcollected"}
outfile_gff_longestcollected_nochr_workaround <-
    here::here("out","gff_ncRNAs_longestverifiedmRNAparts_typeworkaround.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version yeastutrgff",
        paste("##date",lubridate::today()),
        "##selected features: ncRNA, CDS, 5'UTRs, 3'UTRs, introns",
        "##workaround for featureCounts putting feature type in attr"),
    path=outfile_gff_longestcollected_nochr_workaround)
write_gff(gff_longestcollected %>% 
              filter(Type != "chromosome") %>%
              mutate(Attr=paste0(Attr,";FType=",Type)) %>%
              mutate(Type="transcript_region"),
          path=outfile_gff_longestcollected_nochr_workaround,
          append=TRUE)
```


## gff with chromosomes only

```{r gff_chromosomesMito,dependson="load_gff"}
gff_chromosomesMito <- 
        gff_SGDall %>%
            filter( Type == "chromosome") %>%
    mutate(Seqid = str_replace(Seqid,"chrmt","chrMito"),
           Attr = str_replace_all(Attr,"chrmt","chrMito"))
    

outfile_gff_chromosomesMito <-
    here::here("out","gff_ChromosomesMito.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version SGD",
        paste("##date",lubridate::today()),
        "##chromosomes only with chrmt replaced by chrMito"),
    path=outfile_gff_chromosomesMito)
write_gff(gff_chromosomesMito,
          path=outfile_gff_chromosomesMito,
          append=TRUE)

```

# Transcript gff with UTRs and CDS

We do this here with a call to spread (sim to pivot_long), and then a sort of hand-coded spread thing by gene, which is awkward.

```{r txgff_longestcollected,dependson=c("gff_longestcollected","txgff_functions")}
txpartlength_mRNA <- 
    gff_longestcollected %>%
    filter(Seqid != "chrmt", Type %in% c("five_prime_UTR","CDS","three_prime_UTR")) %>%
    mutate(Parent=gene_from_attr(Attr,before="Parent=",gene_str = "[\\w-_]+",after="")) %>%
    group_by(Gene,Type,Parent) %>%
    summarise(Length = sum(End-Start+1)) %>%
    spread(key=Type, value=Length)

txgff_longestcollected <- 
    txpartlength_mRNA %>%
    group_by(Gene) %>%
    do(txgff_frompartlength_1(.)) %>%
    mutate(Source="yeastutrgff",Score=NA,Strand="+")

outfile_txgff_longestcollected <-
    here::here("out",
               "transcriptcentric_longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff")
write_lines(
    x=c("##gff-version 3",
        paste("##source-version yeastutrgff ",lubridate::today())
        ),
    path=outfile_txgff_longestcollected)
write_gff(txgff_longestcollected,
          path=outfile_txgff_longestcollected,
          append=TRUE)
```

```{r txgff_abundantcollected,dependson=c("gff_abundantcollected","txgff_functions")}
txpartlength_mRNA <- 
    gff_abundantcollected %>%
    filter(Seqid != "chrmt", Type %in% c("five_prime_UTR","CDS","three_prime_UTR")) %>%
    mutate(Parent=gene_from_attr(Attr,before="Parent=",gene_str = "[\\w-_]+",after="")) %>%
    group_by(Gene,Type,Parent) %>%
    summarise(Length = sum(End-Start+1)) %>%
    spread(key=Type, value=Length)

txgff_abundantcollected <- 
    txpartlength_mRNA %>%
    group_by(Gene) %>%
    do(txgff_frompartlength_1(.)) %>%
    mutate(Source="yeastutrgff",Score=NA,Strand="+")

outfile_txgff_abundantcollected <-
    here::here("out",
               "transcriptcentric_abundant_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff")
write_lines(
    x=c("##gff-version 3",
        paste("##source-version yeastutrgff ",lubridate::today())
        ),
    path=outfile_txgff_abundantcollected)
write_gff(txgff_abundantcollected,
          path=outfile_txgff_abundantcollected,
          append=TRUE)
```


# Test data for riboviz

This last part is to produce small test data files that we can use to specify and test a better transcript-generating code in riboviz, see [riboviz issue #74](https://github.com/riboviz/riboviz/issues/74).

chromosomeI, 130000:152000, excluding a dubious ORF that was more trouble than it is worth to include.

```{r rv_testregion_YAL10,dependson="load_gffs"}
gff_testregion_YAL10 <- gff_SGDall %>%
    filter(Seqid=="chrI",
           Start > 130000, 
           Start < 151100,
           !str_detect(Attr,"YAL004W"))

outfile_gfftestregion_YAL10 <-
    here::here("out","riboviz_test",
               "saccharomyces_cerevisiae_R64-2-1_20150113_YAL10all.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version SGD R64-2-1, 2015-01-13",
        paste("##subset by Edward Wallace ",lubridate::today())
        ),
    path=outfile_gfftestregion_YAL10)
write_gff(gff_testregion_YAL10,
          path=outfile_gfftestregion_YAL10,
          append=TRUE)

# gff_testregion_YAL10fixedext <-
#     gff_testregion_YAL10 %>%
#     filter(Type=="CDS") %>%
#     
# 
# outfile_gfftestregion_YAL10CDSfixedext <-
#     here::here("out","riboviz_test",
#                "saccharomyces_cerevisiae_R64-2-1_20150113_YALCDS.gff")
# write_lines(
#     x=c("##gff-version 3", 
#         "##source-version SGD R64-2-1, 2015-01-13",
#         paste("##subset by Edward Wallace ",lubridate::today())
#         ),
#     path=outfile_gfftestregion_YAL10CDSfixedext)
# write_gff(gff_testregion_YAL10fixedext,
#           path=outfile_gfftestregion_YAL10CDSfixedext,
#           append=TRUE)

```

```{r rv_testregion_YAL10_txgff,dependson="rv_testregion_YAL10"}
txgff_testregion_YAL10 <- 
    gff_testregion_YAL10 %>%
    filter(Type=="CDS") %>%
    mutate(Name=gene_from_attr(Attr,before="Name=",gene_str = "[\\w-_]+",after=";"),
           Parent=gene_from_attr(Attr,before="Parent=",gene_str = "[\\w-_]+",after=";"),
           Length=End-Start+1) %>%
    group_by(Name) %>%
    summarize(Seqid=unique(Parent),
              Source="yeastutrgff",
              Type="CDS",
              Start=19,
              End=sum(Length)+ 18,
              Score=NA,
              Strand="+",
              Phase=0) %>%
    mutate(Attr=paste("Name=",Name))

outfile_txgfftestregion_YAL10 <-
    here::here("out","riboviz_test",
               "YAL10_transcriptsfixed_R64-2-1_left18_right15.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version SGD R64-2-1, 2015-01-13",
        paste("##subset by Edward Wallace, yeastutrgff",lubridate::today())
        ),
    path=outfile_txgfftestregion_YAL10)
write_gff(txgff_testregion_YAL10,
          path=outfile_txgfftestregion_YAL10,
          append=TRUE)
```

```{r rv_testregion_YBL72ish,dependson="gff_abundantcollected"}
gff_testregion_YBL72ish <- gff_abundantcollected %>%
    filter(Seqid=="chrII",
           Start > 45000, 
           End < 89500)

outfile_gfftestregion_YBL72ish <-
    here::here("out","riboviz_test",
               "saccharomyces_cerevisiae_R64-2-1_withUTRs_YBL72ish.gff")
write_lines(
    x=c("##gff-version 3", 
        "##source-version SGD R64-2-1, 2015-01-13",
        paste("##source-version yeastutrgff",lubridate::today()),
        "##subset by Edward Wallace: ncRNA, CDS, 5'UTRs, 3'UTRs, introns"
        ),
    path=outfile_gfftestregion_YBL72ish)
write_gff(gff_testregion_YBL72ish,
          path=outfile_gfftestregion_YBL72ish,
          append=TRUE)

```
